src/test/librbd/mirror/snapshot/test_mock_CreatePrimaryRequest.cc

```diff
@@ -411,6 +475,102 @@ TEST_F(TestMockMirrorSnapshotCreatePrimaryRequest, SuccessUnlinkNoPeer) {
+      cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY, {"uuid"}, "", CEPH_NOSNAP};
+    ns.complete = true;
+    snap_create(mock_image_ctx, ns, "mirror_snap");
+  }
+
+  cls::rbd::MirrorSnapshotNamespace ns_demoted{
+    cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED, {"uuid"}, "",
+    CEPH_NOSNAP};
+  ns_demoted.complete = true;
+  snap_create(mock_image_ctx, ns_demoted, "mirror_snap");
+
+  for (int i = 0; i < 3; i++) {
+    cls::rbd::MirrorSnapshotNamespace ns{
+      cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY, {"uuid"}, "", CEPH_NOSNAP};
+    ns.complete = true;
+    snap_create(mock_image_ctx, ns, "mirror_snap");
+  }
+
+  InSequence seq;
+
+  expect_clone_md_ctx(mock_image_ctx);
+  MockUtils mock_utils;
+  expect_can_create_primary_snapshot(mock_utils, false, false, true);
+  expect_get_mirror_peers(mock_image_ctx,
+                          {{"uuid", cls::rbd::MIRROR_PEER_DIRECTION_TX, "ceph",
+                            "mirror", "mirror uuid"}}, 0);
+  expect_create_snapshot(mock_image_ctx, 0);
+  expect_refresh_image(mock_image_ctx, 0);
+  MockUnlinkPeerRequest mock_unlink_peer_request;
+  auto it = mock_image_ctx.snap_info.rbegin();
+  auto snap_id = it->first;
+  expect_unlink_peer(mock_image_ctx, mock_unlink_peer_request, snap_id, "uuid",
+                     true, true, true, 0);
+
+  C_SaferCond ctx;
+  auto req = new MockCreatePrimaryRequest(&mock_image_ctx, "gid", CEPH_NOSNAP,
+                                          0U, 0U, nullptr, &ctx);
+  req->send();
+  ASSERT_EQ(0, ctx.wait());
+}
+
+TEST_F(TestMockMirrorSnapshotCreatePrimaryRequest, SuccessUnlinkMultipleNoPeer) {
+  REQUIRE_FORMAT_V2();
+
+  librbd::ImageCtx *ictx;
+  ASSERT_EQ(0, open_image(m_image_name, &ictx));
+  ictx->config.set_val("rbd_mirroring_max_mirroring_snapshots", "3");
+
+  MockTestImageCtx mock_image_ctx(*ictx);
+  for (int i = 0; i < 3; i++) {
+    cls::rbd::MirrorSnapshotState states[] = {
+      cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY,
+      cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY_DEMOTED,
+      cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY,
+      cls::rbd::MIRROR_SNAPSHOT_STATE_NON_PRIMARY_DEMOTED};
+    for (auto state : states) {
+      cls::rbd::MirrorSnapshotNamespace ns{state, {}, "", CEPH_NOSNAP};
+      ns.complete = true;
+      snap_create(mock_image_ctx, ns, "mirror_snap");
+    }
+  }
+
+  InSequence seq;
+
+  expect_clone_md_ctx(mock_image_ctx);
+  MockUtils mock_utils;
+  expect_can_create_primary_snapshot(mock_utils, false, false, true);
+  expect_get_mirror_peers(mock_image_ctx,
+                          {{"uuid", cls::rbd::MIRROR_PEER_DIRECTION_TX, "ceph",
+                            "mirror", "mirror uuid"}}, 0);
+  expect_create_snapshot(mock_image_ctx, 0);
+  expect_refresh_image(mock_image_ctx, 0);
+  MockUnlinkPeerRequest mock_unlink_peer_request;
+  for (auto &it : mock_image_ctx.snap_info) {
+    auto snap_id = it.first;
+    expect_unlink_peer(mock_image_ctx, mock_unlink_peer_request, snap_id, "uuid",
```

Comment:
This loop definitely needs adjustment since `MIRROR_SNAPSHOT_STATE_NON_PRIMARY` and `MIRROR_SNAPSHOT_STATE_NON_PRIMARY_DEMOTED` snapshots should not be considered for unlinking from inside `CreatePrimaryRequest`.

Reply:
I've reread and reinterpreted this comment. For this test, it appears we want to create multiple snapshots so we have a representation of all possible states, but only remove the primary ones. Originally, I understood that as to remove all peerless snaps during this patch and that wasn't the intended function here. That was the source of confusion/misinterpretation of discussion above and here.